ocaml/xapi/xapi_ha_vm_failover.ml

```diff
@@ -175,60 +193,452 @@ let order_f (_, vm_rec) =
+    3. Carry out a binpack plan ignoring VM anti-affinity. *)
+let vm_anti_affinity_evacuation_plan ~__context total_hosts hosts vms
+    vm_can_boot_on_host =
   let config =
     {Binpack.hosts; vms; placement= []; total_hosts; num_failures= 1}
   in
   Binpack.check_configuration config ;
-  debug "Planning configuration for offline agile VMs = %s"
-    (Binpack.string_of_configuration
-       (fun x ->
-         Printf.sprintf "%s (%s)" (Ref.short_string_of x)
-           (Db.Host.get_hostname ~__context ~self:x)
-       )
-       (fun x ->
-         Printf.sprintf "%s (%s)" (Ref.short_string_of x)
-           (Db.VM.get_name_label ~__context ~self:x)
-       )
-       config
-    ) ;
+
+  let vms_increasing = vms |> List.sort bigger_than in
+  let grp_ranked_hosts = anti_affinity_grp_ranked_hosts ~__context hosts in
+
+  let binpack_plan ~__context config vms =
+    debug "Binpack planning configuration = %s"
+      (Binpack.string_of_configuration
+         (fun x ->
+           Printf.sprintf "%s (%s)" (Ref.short_string_of x)
+             (Db.Host.get_name_label ~__context ~self:x)
+         )
+         (fun x ->
+           Printf.sprintf "%s (%s)" (Ref.short_string_of x)
+             (Db.VM.get_name_label ~__context ~self:x)
+         )
+         config
+      ) ;
+    debug "VMs to attempt to evacuate: [ %s ]"
+      (String.concat "; "
+         (vms
+         |> List.map (fun (self, _) ->
+                Printf.sprintf "%s (%s)" (Ref.short_string_of self)
+                  (Db.VM.get_name_label ~__context ~self)
+            )
+         )
+      ) ;
+    let h = Binpack.choose_heuristic config in
+    h.Binpack.get_specific_plan config (List.map fst vms)
+  in
+
+  let try_to_binpack_with_plan_applied_first plan =
+    debug "Try to binpack for the rest VMs" ;
+    let config_after_plan_applied = Binpack.apply_plan config plan in
+    let vms_to_plan =
+      let planned_vm = List.map fst plan in
+      vms |> List.filter (fun (vm, _) -> not (List.mem vm planned_vm))
+    in
+    let config_after_plan_applied =
+      {config_after_plan_applied with vms= vms_to_plan}
+    in
+    let b_plan =
+      binpack_plan ~__context config_after_plan_applied vms_to_plan
+    in
+    match List.length b_plan = List.length vms_to_plan with
+    | true -> (
+        debug "Got final plan with plan applied first." ;
+        match
+          List.for_all (fun (vm, host) -> vm_can_boot_on_host vm host) b_plan
+        with
+        | true ->
+            Some (plan @ b_plan)
+        | false ->
+            debug
+              "Failed to get final plan as the plan from binpack for the rest \
+               VMs is not valid: some VM can't boot on the planned host." ;
+            None
+      )
+    | false ->
+        debug
+          "Failed to get final plan as failed to binpack for all the rest VMs." ;
+        None
+  in
+
+  let rec plan_in_steps ~__context plan_steps =
```

Comment:
What I learned recently :grinning:
```
let plan_with_binpack config plan =
  match config, plan with
  | x, y when x = "x" && y = "a" -> Some (y^x)
  | x, y when x = "x" && y = "b" -> Some (y^x)
  | x, y when x = "x" && y = "x" -> Some (x^y)
  | _ -> None

let plan_for_breadth_first config =
  let plan_for_breadth_first' config =
    match config with
    | x when x = "a" -> Some x
    | _ -> None
  in
  let (let*) = Option.bind in
  let* plan = plan_for_breadth_first' config in
  let* final_plan = plan_with_binpack "x" plan in
  Some final_plan

let plan_for_single_host_failure config =
  let plan_for_single_host_failure' config =
    match config with
    | x when x = "b" -> Some x
    | _ -> None
  in
  let (let*) = Option.bind in
  let* plan = plan_for_single_host_failure' config in
  let* final_plan = plan_with_binpack "x" "b" in
  Some final_plan

let planning config =
  (* a bind for continuation on None *)
  let bind o f = match o with None -> f None | x -> x in
  let ( let* ) = bind in
  let* plan = plan_for_breadth_first config in
  let* plan = plan_for_single_host_failure config in
  let* plan = plan_with_binpack "x" config in
  plan

let () = match planning "y"  with
  | Some plan -> Printf.printf "plan = %s\n" plan
  | None -> Printf.printf "no plan\n"
```

Reply:
I think the current implementation is short, easy to understand and maintainable, is it acceptable?